flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetInputFormat.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.core.fs.FileInputSplit;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.formats.parquet.utils.ParquetRecordReader;
+import org.apache.flink.formats.parquet.utils.ParquetSchemaConverter;
+import org.apache.flink.formats.parquet.utils.RowReadSupport;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.parquet.ParquetReadOptions;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.util.HadoopInputFile;
+import org.apache.parquet.io.InputFile;
+import org.apache.parquet.schema.MessageType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * The base InputFormat class to read from Parquet files.
+ * For specific return types the {@link #convert(Row)} method need to be implemented.
+ *
+ * <P>Using {@link ParquetRecordReader} to Read files instead of {@link org.apache.flink.core.fs.FSDataInputStream},
+ * we override {@link #open(FileInputSplit)} and {@link #close()} to change the behaviors.
+ *
+ * @param <E> The type of record to read.
+ */
+public abstract class ParquetInputFormat<E> extends FileInputFormat<E> implements
+	CheckpointableInputFormat<FileInputSplit, Tuple2<Long, Long>> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(ParquetInputFormat.class);
+
+	private transient Counter recordConsumed;
+
+	protected RowTypeInfo readType;
+
+	protected final TypeInformation[] fieldTypes;
+
+	protected final String[] fieldNames;
+
+	protected transient ParquetRecordReader<Row> parquetRecordReader;
+
+	protected transient long recordsReadSinceLastSync;
+
+	protected long lastSyncedBlock = -1L;
+
+	protected ParquetInputFormat(
+		Path path, TypeInformation[] fieldTypes, String[] fieldNames) {
+		super(path);
+		this.readType = new RowTypeInfo(fieldTypes, fieldNames);
+		this.fieldTypes = readType.getFieldTypes();
+		this.fieldNames = readType.getFieldNames();
+		this.unsplittable = true;
+	}
+
+	@Override
+	public Tuple2<Long, Long> getCurrentState() {
+		return new Tuple2<>(this.lastSyncedBlock, this.recordsReadSinceLastSync);
+	}
+
+	@Override
+	public void open(FileInputSplit split) throws IOException {
+		org.apache.hadoop.conf.Configuration configuration = new org.apache.hadoop.conf.Configuration();
+		InputFile inputFile =
+			HadoopInputFile.fromPath(new org.apache.hadoop.fs.Path(split.getPath().toUri()), configuration);
+		ParquetReadOptions options = ParquetReadOptions.builder().build();
+		ParquetFileReader fileReader = new ParquetFileReader(inputFile, options);
+		MessageType schema = fileReader.getFileMetaData().getSchema();
+		checkSchema(schema);
+		MessageType readSchema = ParquetSchemaConverter.toParquetType(readType);
```

Comment:
readSchema does not depend on anything in `open(...)`, probably can move to constructor

Reply:
The checkSchema function needs to read the parquet footer first. It is too heavy to open and read through all of the files in constructor.